### PR TITLE
Implement DHT Census

### PR DESCRIPTION
**Pull request by pipermerriam**
Replaces #109

This implements full DHT census taking.  The current algorithm is simplistic, effective, and crude.

- perform a RFN on a random location in the network to gather a small set of "known nodes"
- for each node, check liviliness via ping and record `data_radius` value
- for each node, enumerate it's routing table with many FIND_NODES requests on buckets 245-256
- repeat until all nodes have either been verified or time'd out.

This is then saved into the database, with the census itself having simple fields for time bounds, `started_at` and `duration` and each node records (`enr_fk`, `timestamp`, `data_radius`).

No web views have been added for this model yet.

### CODE REVIEW

**Comment by pipermerriam** [comment]
This is ready for review

File: glados-cartographer/src/lib.rs [hunk 1]
```diff
@@ -95,31 +221,247 @@ async fn perform_dht_probe(config: &CartographerConfig, conn: &DatabaseConnectio
...
+
+    // Save record to database
+    let record_model = match record::get_or_create(&enr, &conn).await {
+        Ok(record_model) => {
+            debug!(enr.base64 = enr.to_base64(), "Saved ENR");
+            record_model
+        }
+        Err(err) => {
+            error!(enr.node_id=?H256::from(enr.node_id().raw()), err=?err, "Error saving ENR to database");
+            return;
```
Comment (pipermerriam) [inline glados-cartographer/src/lib.rs:375]: need to drop permit here too.  it'd be nice to have an idiotproof pattern for this, which suggests wrapping this task in one extra layer... will consider.

Comment (pipermerriam) [reply glados-cartographer/src/lib.rs:377]: The cleanup for this was pretty easy.  Curious if there are better idioms for this.

**Comment by perama-v** [comment]
Looks great. I've given it a run and found one issue. 

The second round (15mins/900s later) does not begin if the first round is not `.done()` .

The first round is done if https://github.com/ethereum/glados/blob/edc93c49c53fb78bbd0a2b7fe05be78d94e61b60/glados-cartographer/src/lib.rs#L147 

After 20 mins, the following is observed:

```
[INFO  glados_cartographer] Census progress known=335 alive=47 finished=46 errored=30 pending=259 elapsed=1079 rps=0
```
The census is not done because:
```rs
30 + 46 = 76 != 335 
# (err + fin != known)
```
Thus, many known nodes that are neither finished nor pending (that is, they are `pending`).

I note a few dozen:
```
Error fetching routing table info enr.node_id=<id> msg=Call(Custom(ErrorObject { code: ServerError(-32000), message: "FindNodes request timeout: Timeout", data: None }))
```
I tried adding these as an errored node https://github.com/ethereum/glados/blob/edc93c49c53fb78bbd0a2b7fe05be78d94e61b60/glados-cartographer/src/lib.rs#L445
```rs
census.add_errored(NodeId(enr.node_id().raw())).await; # new
``` 
Thought this was insufficient to resolve the issue. 

One thought was that the second round commencement is gated by the semaphore (default n=1), which
is held by the first round. But increasing concurrency did not result in additional rounds being started while the first was in progress. 

Will think more on it.

File: glados-cartographer/src/lib.rs [hunk 2]
```diff
@@ -95,31 +221,242 @@ async fn perform_dht_probe(config: &CartographerConfig, conn: &DatabaseConnectio
...
+        target.node_id=?H256::from(target.raw()),
+        "Starting DHT census",
     );
 
-    let found_enrs = client
+    // Initialize our search with a random-ish set of ENRs
+    let initial_enrs = client
         .recursive_find_nodes(EthPortalNodeId(target.raw()))
         .await
         .unwrap();
```
Comment (perama-v) [inline glados-cartographer/src/lib.rs:241]: Encountered this unwrap in the wild at 1241 seconds (20 mins). Logs:
```
[INFO  glados_cartographer] Census complete known=420 alive=172 finished=163 errored=257 pending=0 duration=1241 rps=0
...
[DEBUG glados_cartographer] Waiting for channels to exit
...
[INFO  glados_cartographer] Census finished
```
Err:
```
panicked at 'called `Result::unwrap()` on an `Err` value: RequestTimeout'
```

**Comment by pipermerriam** [comment]
@perama-v thanks for the evaluation.  Highlighted that I need to start applying the error handling idioms I've been picking up more consistently.  I've gone through that code with another pass to put in explicit handling of error cases.  Wondering if you would be willing to give the `glados-cartographer/src/lib.rs` one more pass and let me know if you see any patterns that I should change or avoid.  Currently running it for a bit here to see if it more reliably handled errors without fully failing or stalling.

**Comment by pipermerriam** [comment]
ick, still hanging sometimes and I don't know why, no urgency in reviewing this at the moment.

**Comment by perama-v** [comment]
Nice, looks like you fixed it. The problem was an outdated `ethportal-api`, fixed with this:

https://github.com/pipermerriam/glados/pull/2

**Comment by perama-v** [comment]
Some numbers: `trin` observed with ~30 concurrent RPC API connections. Audit finishes after ~1min with alive/errored as follows:
```
Census complete known=433 alive=173 finished=173 errored=260 pending=0 duration=64 rps=9
```

**Review by mrferris — COMMENTED** [review]
Code looks great 👍

Functionality wise, some issues that you may already be aware of:

1. When enumerating a node's routing table, I'm sometimes seeing this: 
`[2023-05-11T19:59:38Z WARN  glados_cartographer] Error fetching routing table info enr.node_id=0x9464e792a5fcef33f027bed8d727422863f7eb185e3a2444358d5275527ce11d distance=245 msg=ParseError(Error("invalid length 0, expected struct FindNodesInfo with 2 elements", line: 1, column: 2))
[2023-05-11T19:59:39Z WARN  glados_cartographer] Error fetching routing table info enr.node_id=0x9464e792a5fcef33f027bed8d727422863f7eb185e3a2444358d5275527ce11d distance=246 msg=ParseError(Error("invalid length 0, expected struct FindNodesInfo with 2 elements", line: 1, column: 2))
[2023-05-11T19:59:39Z WARN  glados_cartographer] Error fetching routing table info enr.node_id=0x9464e792a5fcef33f027bed8d727422863f7eb185e3a2444358d5275527ce11d distance=247 msg=ParseError(Error("invalid length 0, expected struct FindNodesInfo with 2 elements", line: 1, column: 2))
[2023-05-11T19:59:39Z WARN  glados_cartographer] Error fetching routing table info enr.node_id=0x9464e792a5fcef33f027bed8d727422863f7eb185e3a2444358d5275527ce11d distance=248 msg=ParseError(Error("invalid type: string \"enr:-JS4QKcu_bIQ1VbcWhmjmQCGZdeYFovxBamCfW3EzoLOMOKKc_lgFTMZEO1gJk1Qkg4mN38sQrs-cp2g-u5vKCHbA_YBY450IDAuMS4wLTdhZjliYYJpZIJ2NIJpcISfQQZMiXNlY3AyNTZrMaEDoiog6Ml0UGOAkwKKo2FcMaIVIDjYV8PdeNg2fdMgmmGDdWRwgiMo\", expected u8", line: 1, column: 207))
[2023-05-11T19:59:39Z WARN  glados_cartographer] Error fetching routing table info enr.node_id=0x9464e792a5fcef33f027bed8d727422863f7eb185e3a2444358d5275527ce11d distance=249 msg=ParseError(Error("invalid type: string \"enr:-IS4QD8iu2utC1BN59pCxcXWBBAa0QHktQDLeQoWkuAQfUDOTdHfBmsS7_2PwFFouE8DOzXp6o56Oi3GNTHHMafeJ5kBgmlkgnY0gmlwhEFsKgOJc2VjcDI1NmsxoQOjovWhKV_t2veuh2VQ26-p97PuaIdUauvhBV9OhPhCIYN1ZHCCI5M\", expected u8", line: 1, column: 186))
[2023-05-11T19:59:39Z WARN  glados_cartographer] Error fetching routing table info enr.node_id=0x9464e792a5fcef33f027bed8d727422863f7eb185e3a2444358d5275527ce11d distance=250 msg=ParseError(Error("invalid length 0, expected struct FindNodesInfo with 2 elements", line: 1, column: 2))
[2023-05-11T19:59:39Z WARN  glados_cartographer] Error fetching routing table info enr.node_id=0x9464e792a5fcef33f027bed8d727422863f7eb185e3a2444358d5275527ce11d distance=251 msg=ParseError(Error("invalid type: string \"enr:-IS4QIGPVbMufYu4YQfpptnKwUcEoXtwrSji1BNQ09ulPNpAPC6ApQ1HodzrTtFxj6nB0OLa1rF5ozaQSKWOuuy9at4BgmlkgnY0gmlwhEFsKgOJc2VjcDI1NmsxoQLDWHckmer8nmmdYz6kcl5f-mt5NfPT14tH72xvDjR0g4N1ZHCCI6E\", expected u8", line: 1, column: 186))
[2023-05-11T19:59:39Z WARN  glados_cartographer] Error fetching routing table info enr.node_id=0x9464e792a5fcef33f027bed8d727422863f7eb185e3a2444358d5275527ce11d distance=252 msg=ParseError(Error("invalid type: string \"enr:-IS4QHA1PJCdmESyKkQsBmMUhSkRDgwKjwTtPZYMcbMiqCb8I1Xt-Xyh9Nj0yWeIN4S3sOpP9nxI6qCCR1Nf4LjY0IABgmlkgnY0gmlwhEFsKq6Jc2VjcDI1NmsxoQLMWRNAgXVdGc0Ij9RZCPsIyrrL67eYfE9PPwqwRvmZooN1ZHCCI44\", expected u8", line: 1, column: 186))
[2023-05-11T19:59:40Z WARN  glados_cartographer] Error fetching routing table info enr.node_id=0x9464e792a5fcef33f027bed8d727422863f7eb185e3a2444358d5275527ce11d distance=253 msg=ParseError(Error("invalid type: string \"enr:-IS4QDmfFkJzTaS0hFuSrX_Y3VtmkqFAbWeOjoXrh7WlWhptKDZ9FxyfDRZ4YrqKqCWVaCcTDhkOb-HrfwfvAMSLCP4BgmlkgnY0gmlwhEFsKq6Jc2VjcDI1NmsxoQPyTOEJeahGqWxWdRznoJTDYdc24njZ45QMl-EKgMpZ-oN1ZHCCI6Y\", expected u8", line: 1, column: 186))
[2023-05-11T19:59:40Z WARN  glados_cartographer] Error fetching routing table info enr.node_id=0x9464e792a5fcef33f027bed8d727422863f7eb185e3a2444358d5275527ce11d distance=254 msg=ParseError(Error("invalid type: string \"enr:-JS4QHPxpn1TPGhIWf4kPc_M-Ly3ufgyrlMYAOo8bKKeiYcCAtxlJDBKpDvlbbYLmKNp6fAQ--yhiPwR07R76eu2OIYBY450IDAuMS4wLTdlNmNlYYJpZIJ2NIJpcISygPmeiXNlY3AyNTZrMaECu_BN0iAqm75H80HsACpUH4nGKYYcyXQ5doMQT4JwqUGDdWRwgiMo\", expected u8", line: 1, column: 207))
[2023-05-11T19:59:40Z WARN  glados_cartographer] Error fetching routing table info enr.node_id=0x9464e792a5fcef33f027bed8d727422863f7eb185e3a2444358d5275527ce11d distance=255 msg=ParseError(Error("invalid type: string \"enr:-JS4QKJHGuFhaoOic8eo55ZSwpSaZJHZBBgDYs4wMeP9mSFGbmyFi8Z50zrUbwW76iPapT36NQDDm8y46vDuTXPDKmUBY450IDAuMS4wLWU1NzhhMoJpZIJ2NIJpcISAx2MQiXNlY3AyNTZrMaEC2Ri_buS1R8YCshiPb6tBzrAles5zbc3yzEdyGPIFBQiDdWRwgiMo\", expected u8", line: 1, column: 207))
[2023-05-11T19:59:40Z WARN  glados_cartographer] Error fetching routing table info enr.node_id=0x9464e792a5fcef33f027bed8d727422863f7eb185e3a2444358d5275527ce11d distance=256 msg=ParseError(Error("invalid type: string \"enr:-JS4QBpJ-V5ev53L_MoBDuY88pJAPV9yuxcmzEHKk0jS88eJfjFu2qQiGEBbWxxqHvBRJXfyPp2CQssrVMZ8LxVeWAIBY450IDAuMS4wLTdlNmNlYYJpZIJ2NIJpcISygPXviXNlY3AyNTZrMaEC4s_J480xar-qRyEN0rwvo913nVBvKAXbIOoRX9r2nzyDdWRwgiMo\", expected u8", line: 1, column: 207))`



2. I'm seeing much smaller census numbers than you & perama:
`Census complete known=13 alive=10 finished=10 errored=3 pending=0 duration=53 rps=0`
This is running against a trin node whose `portal_historyRoutingTableInfo` endpoint reports as having 71 nodes in its routing table.

**Comment by perama-v** [comment]
Number 1 I encountered, which was the impetus for the bump for from ethportal-api 1.6 to 2.0-alpha

**Comment by perama-v** [comment]
It seems sensitive to concurrency:

- `concurrency=1`: Hangs, with pending=1, never starts second census
- `concurrency=4`: Hangs, with pending=266, never starts second census
- `concurrency=30`: Finishes, starts next census

Which feels like:
- `1`: Too low to add new nodes, too low to finish list of pending nodes.
- `4`: Enough to investigate new nodes, too low to finish list of pending nodes.
- `30`: Enough to investigate new nodes, enough to finish list of pending nodes.

**Comment by pipermerriam** [comment]
I'm pretty sure that this will deadlock at any concurrency level if you let it run long enough....

**Comment by perama-v** [comment]
Fair. I just killed it after about 15 successes and haven't looked too closely for a cause

**Comment by mrferris** [comment]
It looks like the deadlock here was happening when the `to_ping` buffer gets full: the `enumeration` tasks would block while waiting for entries to be taken off the `to_ping` buffer, never freeing up their permits to allow the `liveliness` tasks to take anything off of the `to_ping` buffer. 

I fixed this by creating two semaphores, one for the `enumeration` tasks and one for the `liveliness` tasks. That way they can never block each other. Other solution proposals welcome.

The RPC client was also defaulting to 60 second timeouts, which makes sense for a RFN call, but not for PINGs and FINDNODES. The timeouts for those was lowered to 2 seconds, which speeds things up.

**Pull request merged**